src/bootloader/bootloader.cpp

```diff
@@ -0,0 +1,286 @@
+            fatal_error("bootloader update: failed to seek sector");
+        }
+
+        // erase the sector
+        HAL_FLASH_Unlock();
+        if (!flash_erase_sector(sector)) {
+            fatal_error("bootloader update: failed to erase sector");
+        }
+
+        // program the sector
+        HAL_FLASH_Unlock();
+        bool flash_successful = flash_program_sector(sector, bootloader_bin, [&](size_t bytes_written) {
+            if (sector != 0) {
+                // do not report progress for sector 0, as updating preboot is potentially dangerous
+                // and we want to be as quick as possible and minimize the code running in-between
+                size_t total_written = bytes_in_preceding_sectors + bytes_written;
+                int percent_done = 100 * total_written / total_bytes;
+                progress(percent_done);
+            }
+        });
+        HAL_FLASH_Lock();
+
+        if (!flash_successful) {
+            fatal_error("bootloader update: failed to flash sector");
+        }
+
+        log(LOG_SEVERITY_INFO, "Sector %i flashed successfully", sector);
+        bytes_in_preceding_sectors += bootloader_sector_get_size(sector);
+    }
+}
+
+Version buddy::bootloader::get_version() {
+    return *bootloader_version;
+}
+
+bool buddy::bootloader::needs_update() {
+    auto current = get_version();
+    auto required = buddy::bootloader::required_version;
+
+    return current.major < required.major
+        || (current.major == required.major && current.minor < required.minor)
+        || (current.major == required.major && current.minor == required.minor && current.patch < required.patch);
+}
+
+void buddy::bootloader::update(ProgressHook progress) {
+    auto calc_percent_done = [](int bootstrap_percent, int update_percent) {
+        return (int)(bootstrap_percent * 0.75 + update_percent * 0.25);
+    };
+
+    // get bootloader.bin to the internal flash
+    bool needs_bootstrap = buddy::resources::has_resources(buddy::resources::revision::bootloader) == false;
+    if (needs_bootstrap) {
+        buddy::resources::bootstrap(buddy::resources::revision::bootloader, [&](int percent_done, buddy::resources::BootstrapStage stage) {
+            switch (stage) {
+            case buddy::resources::BootstrapStage::LookingForBbf:
+                progress(0, UpdateStage::LookingForBbf);
+                break;
+            case buddy::resources::BootstrapStage::PreparingBootstrap:
+            case buddy::resources::BootstrapStage::CopyingFiles:
+                progress(calc_percent_done(percent_done, 0), buddy::bootloader::UpdateStage::PreparingUpdate);
+                break;
+            }
+        });
+    }
+
+    std::unique_ptr<FILE, FileDeleter> bootloader_bin(fopen("/internal/res/bootloader.bin", "rb"));
+    if (bootloader_bin.get() == nullptr) {
+        log(LOG_SEVERITY_CRITICAL, "bootloader.bin failed to fopen() after its bootstrap");
+        fatal_error("bootloader.bin failed to open");
```

Comment:
are any of these fatal_errors recoverable somehow, or do they inevitably lead to a board replacement?

Reply:
This one would not lead to board replacement, because we haven't touch the FLASH yet. 

The one below might (in case the erase failed while changing flash content).